tools/syz-testbed/instance.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2021 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"syscall"
+	"time"
+
+	"github.com/google/syzkaller/pkg/config"
+	"github.com/google/syzkaller/pkg/osutil"
+)
+
+// The essential information about an active instance.
+type Instance struct {
+	Name            string
+	Workdir         string
+	BenchFile       string
+	LogFile         string
+	HTTP            string
+	ExecCommand     string
+	ExecCommandArgs []string
+	stopChannel     chan bool
+}
+
+func (inst *Instance) Run() error {
+	const stopDelay = time.Minute
+
+	logfile, err := os.Create(inst.LogFile)
+	if err != nil {
+		return fmt.Errorf("[%s] failed to create logfile: %s", inst.Name, err)
+	}
+	log.Printf("[%s] starting", inst.Name)
+	cmd := osutil.GraciousCommand(inst.ExecCommand, inst.ExecCommandArgs...)
+	cmd.Stdout = logfile
+	cmd.Stderr = logfile
+
+	complete := make(chan error)
+	go func() {
+		complete <- cmd.Run()
+	}()
+
+	select {
+	case err := <-complete:
+		return fmt.Errorf("[%s] stopped: %s", inst.Name, err)
+	case <-inst.stopChannel:
+		// TODO: handle other OSes?
+		cmd.Process.Signal(syscall.SIGTERM)
+		select {
+		case <-complete:
+			// The manager has exited.
+		case <-time.After(stopDelay):
+			// The manager did not exit - kill it.
+			log.Printf("[%s] instance did not exit itself, killing it", inst.Name)
```

Comment:
I don't see where we are actually killing it. We need to wait for the process to finish on every return path.